File: windows/Studio.Core/PadProtocol.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Lernheft.Studio;

/// <summary>
/// Die Sprache zwischen Lernheft Studio (Surface) und Lernheft Pad (iPad).
///
/// Jede Nachricht ist ein JSON-Objekt mit dem Feld „t" für die Art. Die Verbindung ist ein
/// WebSocket im eigenen WLAN. Dieselben Namen stehen in der iPad-App (PadProtocol.swift).
/// </summary>
public static class PadProtocol
{
    public const int Version = 1;

    /// <summary>Hier lauscht das Surface.</summary>
    public const int StudioPort = 47650;

    /// <summary>Hier lauscht das iPad – für den Fall, dass die Windows-Firewall eingehende Verbindungen sperrt.</summary>
    public const int PadPort = 47660;

    public const string Scheme = "lernheftpad";

    // Anmeldung
    public const string Hello = "hello";        // Pad → Studio
    public const string Welcome = "welcome";    // Studio → Pad
    public const string Denied = "denied";      // Studio → Pad
    public const string Ping = "ping";
    public const string Pong = "pong";
    public const string Bye = "bye";

    // Die offene Notiz (Studio → Pad)
    public const string Note = "note";          // Kopfdaten oder „keine Notiz offen"
    public const string Ink = "ink";            // alle Striche
    public const string Images = "images";      // eingefügte Bilder
    public const string Text = "text";          // Textebene einer Seite als Bild
    public const string View = "view";          // Ausschnitt, den das Surface gerade zeigt

    // Zeichnen
    public const string Live = "live";          // Pad → Studio: Strich, während er entsteht
    public const string Ops = "ops";            // beide Richtungen: Striche hinzu / weg

    // Wünsche vom iPad
    public const string Pages = "pages";        // mehr Seiten
    public const string EraseImage = "eraseImage";
    public const string AddText = "addText";
    public const string DeleteIn = "deleteIn";
    public const string Convert = "convert";    // Handschrift → Text/Schönschrift
    public const string Converted = "converted";
    public const string Ai = "ai";              // Ausschnitt an den KI-Helfer
    public const string Open = "open";          // Notiz am Surface öffnen
    public const string NewNote = "newNote";
    public const string Theme = "theme";
    public const string Toast = "toast";        // kurze Meldung fürs iPad

    // Umzug der alten Notizen vom iPad
    public const string ImportBegin = "importBegin";
    public const string ImportFile = "importFile";
    public const string ImportEnd = "importEnd";
    public const string ImportDone = "importDone";

    public static JsonObject Message(string type) => new() { ["t"] = type };

    public static string TypeOf(JsonObject message) => message["t"]?.GetValue<string>() ?? "";

    public static string? String(this JsonObject message, string key) =>
        message[key] is JsonValue value && value.TryGetValue<string>(out var text) ? text : null;

    public static double Number(this JsonObject message, string key, double fallback = 0)
    {
        if (message[key] is not JsonValue value) return fallback;
        if (value.TryGetValue<double>(out var number)) return number;
        if (value.TryGetValue<int>(out var integer)) return integer;
        if (value.TryGetValue<long>(out var big)) return big;
        return fallback;
    }

    public static bool Flag(this JsonObject message, string key) =>
        message[key] is JsonValue value && value.TryGetValue<bool>(out var flag) && flag;

    public static Guid? Guid(this JsonObject message, string key) =>
        System.Guid.TryParse(message.String(key), out var id) ? id : null;

    public static JsonArray StrokesToJson(IEnumerable<InkStroke> strokes)
    {
        var array = new JsonArray();
        foreach (var stroke in strokes) array.Add(JsonSerializer.SerializeToNode(stroke, InkDocument.Options));
        return array;
    }

    public static List<InkStroke> StrokesFromJson(JsonNode? node)
    {
        if (node is not JsonArray array) return new List<InkStroke>();
        var result = new List<InkStroke>();
        foreach (var item in array)
        {
            if (item is null) continue;
            var stroke = item.Deserialize<InkStroke>(InkDocument.Options);
            if (stroke is not null && stroke.Count > 0) result.Add(stroke);
        }
        return result;
    }

    public static List<double> NumbersFromJson(JsonNode? node)
    {
        var result = new List<double>();
        if (node is not JsonArray array) return result;
        foreach (var item in array)
        {
            if (item is JsonValue value && value.TryGetValue<double>(out var number)) result.Add(number);
        }
        return result;
    }

    public static List<string> StringsFromJson(JsonNode? node)
    {
        var result = new List<string>();
        if (node is not JsonArray array) return result;
        foreach (var item in array)
        {
            if (item is JsonValue value && value.TryGetValue<string>(out var text)) result.Add(text);
        }
        return result;
    }
}

/// <summary>Ein gekoppeltes iPad – es darf sich später ohne neuen Code wieder melden.</summary>
public class PairedDevice
{
    public string DeviceId { get; set; } = "";
    public string Name { get; set; } = "iPad";
    public string PairKey { get; set; } = "";
    public double LastSeen { get; set; } = AppleTime.Now;
    public string LastAddress { get; set; } = "";
}

/// <summary>
/// Ein Kopplungsfenster: ein sechsstelliger Code zum Abtippen und ein langes Geheimnis im QR-Code.
/// Beides gilt nur, solange das Fenster offen ist.
/// </summary>
public class PairingSession
{
    public string Code { get; }
    public string Secret { get; }
    public DateTime Expires { get; }
    private int _failures;

    public PairingSession(TimeSpan lifetime)
    {
        Code = RandomNumberGenerator.GetInt32(0, 1_000_000).ToString("000000", CultureInfo.InvariantCulture);
        Secret = RandomToken(18);
        Expires = DateTime.UtcNow + lifetime;
    }

    public bool IsValid => DateTime.UtcNow < Expires && _failures < 8;

    public bool Accepts(string? code, string? secret)
    {
        if (!IsValid) return false;
        var ok = (!string.IsNullOrEmpty(secret) && CryptographicOperations.FixedTimeEquals(
                     System.Text.Encoding.UTF8.GetBytes(secret), System.Text.Encoding.UTF8.GetBytes(Secret)))
                 || (!string.IsNullOrEmpty(code) && Normalize(code) == Code);
        if (!ok) _failures++;
        return ok;
    }

    public static string Normalize(string code) => new(code.Where(char.IsDigit).ToArray());

    public string FormattedCode => Code[..3] + " " + Code[3..];

    /// <summary>Inhalt des QR-Codes: alles, was das iPad zum Verbinden braucht.</summary>
    public string QrPayload(string serverId, string serverName, IEnumerable<string> addresses, int port) =>
        $"{PadProtocol.Scheme}://pair?v={PadProtocol.Version}" +
        $"&id={Uri.EscapeDataString(serverId)}" +
        $"&n={Uri.EscapeDataString(serverName)}" +
        $"&h={Uri.EscapeDataString(string.Join(",", addresses))}" +
        $"&p={port}" +
        $"&s={Uri.EscapeDataString(Secret)}";

    public static string RandomToken(int bytes)
    {
        var data = RandomNumberGenerator.GetBytes(bytes);
        return System.Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');
    }
}

File: ipad/LernheftPad/PadProtocol.swift
import Foundation

/// Die Sprache zwischen Lernheft Stift (iPad) und Lernheft Studio (Surface).
/// Jede Nachricht ist ein JSON-Objekt mit dem Feld „t“ für die Art. Dieselben Namen stehen
/// in Lernheft Studio (PadProtocol.cs).
enum PadProtocol {
    static let version = 1
    static let studioPort: UInt16 = 47650
    static let padPort: UInt16 = 47660
    static let scheme = "lernheftpad"
    static let studioService = "_lernheft._tcp"
    static let padService = "_lernheftpad._tcp"

    // Anmeldung
    static let hello = "hello"
    static let welcome = "welcome"
    static let denied = "denied"
    static let ping = "ping"
    static let pong = "pong"
    static let bye = "bye"

    // Die offene Notiz (Surface → iPad)
    static let note = "note"
    static let ink = "ink"
    static let images = "images"
    static let text = "text"
    static let view = "view"

    // Zeichnen
    static let live = "live"
    static let ops = "ops"

    // Wünsche vom iPad
    static let pages = "pages"
    static let eraseImage = "eraseImage"
    static let addText = "addText"
    static let deleteIn = "deleteIn"
    static let convert = "convert"
    static let converted = "converted"
    static let ai = "ai"
    static let open = "open"
    static let newNote = "newNote"
    static let theme = "theme"
    static let toast = "toast"

    // Umzug der alten Notizen
    static let importBegin = "importBegin"
    static let importFile = "importFile"
    static let importEnd = "importEnd"
    static let importDone = "importDone"

    static func message(_ type: String) -> JSON { ["t": type] }

    /// Schreibt Zahlen so kurz wie möglich („102.15“ statt „102.15000000000001“) – Striche
    /// bestehen aus Tausenden Zahlen, das spart gut die Hälfte der Übertragung.
    static func encode(_ message: JSON) -> Data? {
        var output = ""
        output.reserveCapacity(256)
        guard JSONWriter.write(message, into: &output) else { return nil }
        return output.data(using: .utf8)
    }

    static func decode(_ data: Data) -> JSON? {
        (try? JSONSerialization.jsonObject(with: data)) as? JSON
    }
}

typealias JSON = [String: Any]

extension Dictionary where Key == String, Value == Any {
    var messageType: String { self["t"] as? String ?? "" }

    func string(_ key: String) -> String? { self[key] as? String }

    func number(_ key: String, _ fallback: Double = 0) -> Double {
        (self[key] as? NSNumber)?.doubleValue ?? fallback
    }

    func flag(_ key: String) -> Bool {
        if let value = self[key] as? Bool { return value }
        return (self[key] as? NSNumber)?.boolValue ?? false
    }

    func objects(_ key: String) -> [JSON] { self[key] as? [JSON] ?? [] }

    func strings(_ key: String) -> [String] { self[key] as? [String] ?? [] }

    func numbers(_ key: String) -> [Double] {
        (self[key] as? [Any])?.compactMap { ($0 as? NSNumber)?.doubleValue } ?? []
    }
}

/// Was im QR-Code am Surface steht: lernheftpad://pair?v=1&id=…&n=…&h=ip1,ip2&p=47650&s=…
struct PairingOffer: Equatable {
    var serverId: String
    var serverName: String
    var hosts: [String]
    var port: UInt16
    var secret: String

    init?(_ text: String) {
        guard let components = URLComponents(string: text.trimmingCharacters(in: .whitespacesAndNewlines)),
              components.scheme == PadProtocol.scheme else { return nil }
        var values: [String: String] = [:]
        for item in components.queryItems ?? [] { values[item.name] = item.value ?? "" }
        guard let id = values["id"], !id.isEmpty, let secret = values["s"], !secret.isEmpty else { return nil }
        serverId = id
        serverName = values["n"] ?? "Surface"
        hosts = (values["h"] ?? "").split(separator: ",").map { String($0) }.filter { !$0.isEmpty }
        port = UInt16(values["p"] ?? "") ?? PadProtocol.studioPort
        self.secret = secret
    }
}

enum JSONWriter {
    static func write(_ value: Any, into output: inout String) -> Bool {
        switch value {
        case let text as String:
            writeString(text, into: &output)
        case let flag as Bool where !(value is NSNumber) || CFGetTypeID(value as CFTypeRef) == CFBooleanGetTypeID():
            output += flag ? "true" : "false"
        case let number as Int:
            output += String(number)
        case let number as Double:
            guard number.isFinite else { return false }
            output += number == number.rounded() && abs(number) < 1e15 ? String(Int64(number)) : String(number)
        case let number as CGFloat:
            return write(Double(number), into: &output)
        case let number as Float:
            return write(Double(number), into: &output)
        case let number as NSNumber:
            return write(number.doubleValue, into: &output)
        case let list as [Any]:
            output += "["
            for (index, item) in list.enumerated() {
                if index > 0 { output += "," }
                guard write(item, into: &output) else { return false }
            }
            output += "]"
        case let object as [String: Any]:
            output += "{"
            var first = true
            for (key, item) in object {
                if !first { output += "," }
                first = false
                writeString(key, into: &output)
                output += ":"
                guard write(item, into: &output) else { return false }
            }
            output += "}"
        case is NSNull:
            output += "null"
        default:
            return false
        }
        return true
    }

    private static func writeString(_ text: String, into output: inout String) {
        output += "\""
        for scalar in text.unicodeScalars {
            switch scalar {
            case "\"": output += "\\\""
            case "\\": output += "\\\\"
            case "\n": output += "\\n"
            case "\r": output += "\\r"
            case "\t": output += "\\t"
            default:
                if scalar.value < 0x20 {
                    output += String(format: "\\u%04X", scalar.value)
                } else {
                    output.unicodeScalars.append(scalar)
                }
            }
        }
        output += "\""
    }
}
